frontends/common/constantFolding.cpp

```diff
@@ -440,6 +441,15 @@ const IR::Node *DoConstantFolding::compare(const IR::Operation_Binary *e) {
         }
         bool bresult = (left->value == right->value) == eqTest;
         return new IR::BoolLiteral(e->srcInfo, IR::Type_Boolean::get(), bresult);
+    } else if (eleft->is<IR::StringLiteral>()) {
```

Comment:
Can this be simplified to:
```c++
} else if (const auto *left = eleft->to<IR::StringLiteral>()) {
..
```

This would skip double cast. The code for `BoolLiteral` likely would need to be fixed as well :)

Reply:
Also, maybe it would help to add comment to `bool eqTest` above to clarify that `compare` is called only for `IR::Equ` and `IR::Neq`, so other cases are not possible. It tool me a while to figure out why other `IR::Operation_Binary` cases are not handled...